pkg/compact/compact.go

```diff
@@ -1007,53 +1014,66 @@ func (cg *Group) compact(ctx context.Context, dir string, planner Planner, comp
 
 	// Once we have a plan we need to download the actual data.
 	begin := time.Now()
+	g, errCtx := errgroup.WithContext(ctx)
+	g.SetLimit(cg.compactionBlocksConcurrency)
 
 	toCompactDirs := make([]string, 0, len(toCompact))
-	for _, meta := range toCompact {
-		bdir := filepath.Join(dir, meta.ULID.String())
-		for _, s := range meta.Compaction.Sources {
+	for _, m := range toCompact {
+		bdir := filepath.Join(dir, m.ULID.String())
+		for _, s := range m.Compaction.Sources {
 			if _, ok := uniqueSources[s]; ok {
 				return false, ulid.ULID{}, halt(errors.Errorf("overlapping sources detected for plan %v", toCompact))
 			}
 			uniqueSources[s] = struct{}{}
 		}
 
-		tracing.DoInSpanWithErr(ctx, "compaction_block_download", func(ctx context.Context) error {
-			err = block.Download(ctx, cg.logger, cg.bkt, meta.ULID, bdir, objstore.WithFetchConcurrency(cg.blockFilesConcurrency))
-			return err
-		}, opentracing.Tags{"block.id": meta.ULID})
-		if err != nil {
-			return false, ulid.ULID{}, retry(errors.Wrapf(err, "download block %s", meta.ULID))
-		}
+		func(ctx context.Context, meta *metadata.Meta) {
+			g.Go(func() error {
+				tracing.DoInSpanWithErr(ctx, "compaction_block_download", func(ctx context.Context) error {
```

Comment:
Hi @alanprot, if I understand correctly, it is still first IO bound -> CPU bound because for each block the process is still sequential. What's the difference of this between your original commit of the previous pr?

Reply:
It is still "Io Bound" -> "CPU Bound".. The advantage here are:
 *  We now can have some blocks downloading while others are reading the index
 * We can use more than 1 CPU to read the indexes now.
 
This change is pretty similar to my original one: https://github.com/thanos-io/thanos/pull/5476/commits/b83620b0308ed37ca96bef11e59775d241155b18

Reply:
The change is almost the same as the original commit but I thought you would change some logic. Maybe I misunderstood it yesterday. Thanks for the clarification! 💯

Reply:
Yes, I agree. Compared to the index verification operation which is CPU bound, the IO part is taking a much longer time to download the block which is a major factor in this case. I think 1 and 2 are pretty similar.